Crash Reporting implementation

SenRamakri opened the pull request:
### Description
This PR implements Crash-reporting feature. Please see the design-document for this feature at https://github.com/ARMmbed/mbed-os/pull/8561

There is also example application for this feature at - https://github.com/ARMmbed/mbed-os-example-crash-reporting. The example application(mbed-os.lib) will be updated to point to right version once this PR is merged.

This PR is targeted for 5.11 release.

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/device/TOOLCHAIN_GCC_ARM/MK66FN2M0xxx18.ld:
```diff
@@ -195,6 +196,18 @@ SECTIONS
+  
+  .crash_data_ram :
+  {
+    . = ALIGN(4);
```
For ARMC6 support in future we have updated all linker scripts to be 8-byte aligned.

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_GCC_ARM/MK64FN1M0xxx12.ld:
```diff
@@ -53,6 +53,16 @@ ENTRY(Reset_Handler)
+ * uVisor is enabled, this is also the maximum available heap size. The
+ * HEAP_SIZE value is set by uVisor porters to balance the size of the legacy
+ * heap and the page heap in uVisor applications. */
+__heap_size__ = 0x6000;
```
Minimum heap size of 0x6000? Seems copy paste error here, as it states uVisor in comments

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_GCC_ARM/MK64FN1M0xxx12.ld:
```diff
@@ -92,22 +93,32 @@ SECTIONS
   {
     __VECTOR_TABLE = .;
-    . = ALIGN(8);
+    . = ALIGN(4);
```
Why was alignment changed from 8-byte to 4-byte?

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_GCC_ARM/MK64FN1M0xxx12.ld:
```diff
@@ -192,15 +203,69 @@ SECTIONS
+    __CRASH_DATA_RAM_END__ = .; /* Define a global symbol at data end */
+  } > m_data
+
+  /* uVisor own memory and private box memories
```
uvisor changes?

SenRamakri replied:
I think these are rebase issues, I'll fix them, Thanks for pointing them out.

kegilbert commented on cmsis/TARGET_CORTEX_M/mbed_fault_handler.c:
```diff
@@ -69,24 +75,24 @@ void mbed_fault_handler (uint32_t fault_type, void *mbed_fault_context_in)
+        mbed_error_printf("\nR%-4d: %08X", i, ((uint32_t *)(mbed_fault_context))[i]);  
     }
         
-    mbed_error_printf("\nSP   : %08" PRIX32
```
Why are the printf format directives being removed? These were a fairly recent addition:

https://github.com/ARMmbed/mbed-os/pull/8601

deepikabhavnani commented on platform/mbed_error.c:
```diff
@@ -188,24 +266,89 @@ WEAK MBED_NORETURN mbed_error_status_t mbed_error(mbed_error_status_t error_stat
+mbed_error_status_t mbed_reset_reboot_error_info()
+{
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    memset(report_error_ctx, 0, sizeof(mbed_error_ctx) ); 
```
mbed_reset_reboot_error_info - Is this thread safe?

SenRamakri replied:
Its currently not, I agree that its better to make it thread-safe to avoid any assumptions on app side. I'll push a fix for this.

deepikabhavnani replied:
I was thinking of adding in docs " Not thread safe" instead of making it thread safe, as we will have more corner cases

SenRamakri replied:
Since the user cannot control when the error may get triggered I think its better to protect this struct anyway.

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_ARM_STD/MK64FN1M0xxx12.sct:
```diff
@@ -64,10 +64,6 @@
   #define MBED_APP_SIZE 0x100000
 #endif
 
-#if !defined(MBED_BOOT_STACK_SIZE)
```
Default stack size is removed by mistake

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_ARM_STD/MK64FN1M0xxx12.sct:
```diff
@@ -80,8 +76,11 @@
+#define m_crash_report_ram_start       (m_interrupts_ram_start + m_interrupts_ram_size)
+#define m_crash_report_ram_size        (0x100)
+
+#define m_data_start                   (m_crash_report_ram_start + m_crash_report_ram_size)
```
```suggestion
#define m_data_start                   (m_crash_report_ram_start + m_crash_report_ram_size + m_interrupts_ram_size)
```

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/device/TOOLCHAIN_ARM_STD/MK66FN2M0xxx18.sct:
```diff
@@ -72,8 +72,11 @@
+#define m_crash_report_ram_start       (m_interrupts_ram_start + m_interrupts_ram_size)
+#define m_crash_report_ram_size        (0x100)
+
+#define m_data_start                   (m_crash_report_ram_start + m_crash_report_ram_size)
```
```suggestion
#define m_data_start                   (m_crash_report_ram_start + m_crash_report_ram_size + m_interrupts_ram_size)
```

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_ARM_STD/MK64FN1M0xxx12.sct:
```diff
@@ -90,7 +89,7 @@
   #define Stack_Size                   __stack_size__
 #else
-  #define Stack_Size                   MBED_BOOT_STACK_SIZE
+  #define Stack_Size                   0x0400
```
```suggestion
  #define Stack_Size                   MBED_BOOT_STACK_SIZE
```

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_ARM_STD/MK64FN1M0xxx12.sct:
```diff
@@ -126,6 +127,4 @@ LR_IROM1 m_interrupts_start m_text_start+m_text_size-m_interrupts_start {   ; lo
   }
   RW_IRAM1 ImageLimit(RW_m_data_2) { ; Heap region growing up
   }
-  ARM_LIB_STACK m_data_2_start+m_data_2_size EMPTY -Stack_Size { ; Stack region growing down
```
ARM_LIB_STACK is removed, could be conflict issue

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_IAR/MK64FN1M0xxx12.icf:
```diff
@@ -57,13 +61,6 @@ if (!isdefinedsymbol(MBED_APP_SIZE)) {
 }
 
-if (!isdefinedsymbol(MBED_BOOT_STACK_SIZE)) {
-    define symbol MBED_BOOT_STACK_SIZE = 0x400;
```
This code should not be removed, conflict issue

SenRamakri committed:
Error and fault handling changes for crash reporting

SenRamakri committed:
Linker script modifications for crash data region

SenRamakri committed:
Enable crash capture feature for platforms with linkerscript changes

SenRamakri committed:
Added greentea test

SenRamakri commented:
@deepikabhavnani and @kegilbert - I have cleaned up the PR, conflict issues you pointed out have been removed. Please review.

deepikabhavnani reviewed (commented):
Changes look good to me, I have one query related to public API's - `mbed_reset_reboot_error_info` and `mbed_reset_reboot_count` - Do we have any limitation on when and where they can be used? Else we need to consider the thread safety here. 

Example: Clearing an error count when error is in progress, during initialization of multiple threads / modules in system could cause issue.

SenRamakri committed:
Add thread safety across reboot error info operations

kegilbert commented on platform/mbed_error.c:
```diff
@@ -44,10 +47,46 @@ static core_util_atomic_flag error_in_progress = CORE_UTIL_ATOMIC_FLAG_INIT;
+    const unsigned int polynomial = 0x04C11DB7; /* divisor is 32bit */
+    unsigned int crc = 0; /* CRC value is 32bit */
+
+    for( ;datalen>=0; datalen-- ) {
```
Minor Style: Space between for and '('

SenRamakri replied:
Fixed

kegilbert commented on platform/mbed_error.c:
```diff
@@ -188,24 +265,97 @@ WEAK MBED_NORETURN mbed_error_status_t mbed_error(mbed_error_status_t error_stat
+    uint32_t crc_val = 0;
+    crc_val = compute_crc32( (unsigned char *)report_error_ctx, ((uint32_t)&(report_error_ctx->crc_error_ctx) - (uint32_t)report_error_ctx) );
+    //Read report_error_ctx and check if CRC is correct for report_error_ctx
+    if(report_error_ctx->crc_error_ctx == crc_val) {
```
Minor Style: Space between `if` and `(`

SenRamakri replied:
Fixed

kegilbert commented on platform/mbed_error.c:
```diff
@@ -188,24 +265,97 @@ WEAK MBED_NORETURN mbed_error_status_t mbed_error(mbed_error_status_t error_stat
+mbed_error_status_t mbed_reset_reboot_count()
+{
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    if(is_reboot_error_valid) {
```
Minor Style: Space between `if` and `(`

kegilbert commented on platform/mbed_error.c:
```diff
@@ -188,24 +265,97 @@ WEAK MBED_NORETURN mbed_error_status_t mbed_error(mbed_error_status_t error_stat
+    mbed_error_status_t status = MBED_ERROR_ITEM_NOT_FOUND;
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    if (is_reboot_error_valid) {
+        if(error_info != NULL) {
```
Minor Style: Space between `if` and `(`

kegilbert commented on platform/mbed_error.c:
```diff
@@ -152,6 +191,44 @@ static mbed_error_status_t handle_error(mbed_error_status_t error_status, unsign
+        
+        //Enforce max-reboot only if auto reboot is enabled
+#if MBED_CONF_PLATFORM_FATAL_ERROR_AUTO_REBOOT_ENABLED        
+        if( report_error_ctx->error_reboot_count > MBED_CONF_PLATFORM_ERROR_REBOOT_MAX ) {
```
Minor Style: Space between `if` and `(`

kegilbert commented on platform/mbed_error.c:
```diff
@@ -152,6 +191,44 @@ static mbed_error_status_t handle_error(mbed_error_status_t error_status, unsign
+    uint32_t crc_val = 0;
+    crc_val = compute_crc32( (unsigned char *)report_error_ctx, ((uint32_t)&(report_error_ctx->crc_error_ctx) - (uint32_t)report_error_ctx) );
+    //Read report_error_ctx and check if CRC is correct for report_error_ctx
+    if((report_error_ctx->crc_error_ctx == crc_val) && (report_error_ctx->is_error_processed == 0)) {
```
Minor Style: Space between `if` and `(`

kegilbert reviewed (approved):
Platform changes LGTM except for some very minor style nits above (didn't check out the linker script changes much).

kjbracey commented on cmsis/TARGET_CORTEX_M/mbed_fault_handler.c:
```diff
@@ -20,17 +20,23 @@
-mbed_fault_context_t mbed_fault_context;
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    //Global for populating the context in exception handler
+    mbed_fault_context_t *mbed_fault_context=(mbed_fault_context_t *)((uint32_t)FAULT_CONTEXT_LOCATION);
```
Pointers should be const to save RAM (`mbed_fault_context * const mbed_fault_context`)

SenRamakri replied:
Fixed

kjbracey commented on cmsis/TARGET_CORTEX_M/mbed_fault_handler.h:
```diff
@@ -55,4 +59,18 @@ typedef struct {
+ * @param  fault_context        Pointer to mbed_fault_context_t struct allocated by the caller. This is the mbed_fault_context_t info captured as part of the fatal exception which triggered the reboot.
+ * @return                      0 or MBED_SUCCESS on success.
+ *                              MBED_ERROR_INVALID_ARGUMENT in case of invalid error_info pointer
+ *                              MBED_ERROR_ITEM_NOT_FOUND if no reboot context is currently captured by teh system 
```
the

kjbracey commented on platform/mbed_error.c:
```diff
@@ -44,10 +47,46 @@ static core_util_atomic_flag error_in_progress = CORE_UTIL_ATOMIC_FLAG_INIT;
+
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    //Global for populating the context in exception handler
+    static mbed_error_ctx *report_error_ctx=(mbed_error_ctx *)((uint32_t)ERROR_CONTEXT_LOCATION);
```
(uint32_t) cast isn't doing anything I can see?

SenRamakri replied:
Fixed

kjbracey commented on platform/mbed_error.c:
```diff
@@ -152,6 +191,44 @@ static mbed_error_status_t handle_error(mbed_error_status_t error_status, unsign
+{
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    uint32_t crc_val = 0;
+    crc_val = compute_crc32( (unsigned char *)report_error_ctx, ((uint32_t)&(report_error_ctx->crc_error_ctx) - (uint32_t)report_error_ctx) );
```
The complicated bit here is  `offsetof(mbed_error_ctx, crc_error_ctx)`, no?

SenRamakri replied:
Ah yes, thanks for pointing that out and offsetof is much easier for readability.

kjbracey commented on platform/mbed_error.c:
```diff
@@ -44,10 +47,46 @@ static core_util_atomic_flag error_in_progress = CORE_UTIL_ATOMIC_FLAG_INIT;
+//we dont have many uses cases to create a C wrapper for MbedCRC and the data
+//we calculate CRC on in this context is very less we will use a local 
+//implementation here.
+static unsigned int compute_crc32(unsigned char *data, int datalen)
```
If this took `const void *` you'd only need one cast inside here, not 1 at every callsite.

SenRamakri replied:
Fixed

kjbracey commented on platform/mbed_error.c:
```diff
@@ -188,24 +265,97 @@ WEAK MBED_NORETURN mbed_error_status_t mbed_error(mbed_error_status_t error_stat
+    //We need not call delete_mbed_crc(crc_obj) here as we are going to reset the system anyway, and calling delete while handling a fatal error may cause nested exception
+#if MBED_CONF_PLATFORM_FATAL_ERROR_AUTO_REBOOT_ENABLED
+    system_reset();//do a system reset to get the system rebooted
+    while(1);
```
while(1) is dead here. system_reset is MBED_NORETURN.

I would be strongly inclined to have an exponential backoff delay on each reboot. 5 seconds first time, say, and double each time.

SenRamakri replied:
while(1) was an attempt to stop the system in case system_reset() ever changed behavior. But I think its confusing, so will remove it.

I initially thought of having delay before reboot, but later I decided it wont help much, unless we want to provide opportunity to attach a debugger or something, is that why we should add the backoff delay?

kjbracey replied:
It's more a stability principle - each time the device boots, it may well be using extra network resources to reinitialise itself.  The problem leading to crash may even be exacerbated by overall system load caused by other rebooting devices. A network could collapse as a result.

A general networking principle is that retries should be exponentially backed off to guarantee that the network is stable, and doesn't undergo congestion collapse. (Many protocols and standards use parameters based on chapter 14 of [RFC 3315](https://tools.ietf.org/html/rfc3315) - initial time, maximum time, maximum count...)

Obviously we would require someone at some point to reset the count, and hence the backoff, but you've left that application dependent. Were that to be reset too readily, there might be a problem.

In this case, maybe you don't want to literally double the "delay after crashing before rebooting", but rather double the "minimum time since we last rebooted". Not sure what your best timer for that would be - can't safely use `Kernel::get_ms_count`, so might need to have a boot-started `LowPowerTimer'? 

If you're initially configuring the count very low though, this is less necessary - you'd want this with high or unlimited counts. So it's maybe an extension.

SenRamakri replied:
Thanks for the explanation and pointers, currently I'm setting the count to 1 for now, so I would leave things as it is. But I would capture this as a note in my documentation about how to enable/configure crash-reporting. Hope that helps. And in future if we are increasing the default count, I'll add something to delay the reboot on every subsequent reboot.

kjbracey commented on platform/mbed_lib.json:
```diff
@@ -110,6 +126,71 @@
+        "DISCO_L475VG_IOT01A": {
+            "crash-capture-enabled": true,
+            "reboot-crash-report-enabled": true,
+            "fatal-error-auto-reboot-enabled": true
```
The reboot is a pretty big default behaviour functional change. Are we sure on that one?

SenRamakri replied:
Yeah its a big behavior change. But since the default config is to reboot only once(error_reboot_max = 1) it will reboot and halt, that way it shouldn't impact any tests or other tools. Do you still think we should keep the auto-reboot disabled, any issues you are seeing?

kjbracey replied:
I'm not seeing any particular issues myself, just seems like it might be a bit unexpected to end users.

SenRamakri replied:
I see that, I would leave that as it is for now unless I get a request to change it explicitly, hope that's ok.

bulislaw replied:
Can we make sure that gets into the release notes, talk to Mohit.

kjbracey commented on platform/mbed_error.c:
```diff
@@ -369,17 +519,19 @@ static void print_error_report(const mbed_error_ctx *ctx, const char *error_msg,
 #if MBED_CONF_PLATFORM_ERROR_ALL_THREADS_INFO && defined(MBED_CONF_RTOS_PRESENT)
-    mbed_error_printf("\nNext:");
-    print_thread(osRtxInfo.thread.run.next);
+    if(print_thread_info == true) {
```
Please don't use `if (boolean == true)`. Just `if (boolean)`.

SenRamakri replied:
Fixed

kjbracey commented on platform/mbed_lib.json:
```diff
@@ -102,6 +102,22 @@
+            "help": "Enables crash context capture when the system enters a fatal error/crash.",
+            "value": false
+        },
+        "error-reboot-max": {
```
It's not clear what 0 means here. Sounds as if it means don't reboot, but it's actually reboot once? 1 would mean reboot twice?

SenRamakri replied:
That's correct, 0 means it still causes 1 reboot. May be its confusing as you see it, its also tricky to document that I guess. I'm going to change that to - 0 means no reboot, 1 means 1 reboot etc. Hope that helps.

SenRamakri commented:
@kegilbert and @kjbracey-arm - I'll be pushing a change to address your comments.

SenRamakri committed:
Change the reboot-max value semantics and code style and other fixes

SenRamakri commented:
@kjbracey-arm - I have pushed changes addressing your comments. I also have couple of questions, please re-review.

studavekar commented:
Build Pipeline is free, The logs are not clear of jenkins-ci/build-ARM for failure. Triggering a build to verify the failure 

/morph build

SenRamakri committed:
Fix linker script errors

cmonr commented:
@SenRamakri Was this just rebased, or were additional changes made to address the build errors?

SenRamakri commented:
@cmonr - Both, we did rebase and also addressed couple of build errors.

SenRamakri commented:
Pushed again to resolve the conflict with mbed_lib.json.

cmonr commented:
/morph build

0xc0170 commented:
Tests will finish soon. I reviewed one failure, looks related to this additions. Please review

kjbracey commented on platform/mbed_error.c:
```diff
@@ -44,10 +47,46 @@ static core_util_atomic_flag error_in_progress = CORE_UTIL_ATOMIC_FLAG_INIT;
+    unsigned int crc = 0; /* CRC value is 32bit */
+
+    for ( ;datalen>=0; datalen-- ) {
+        unsigned char b = (*(unsigned char *)data);
```
You're not advancing the data pointer! Neater to just do

    const unsigned char *ptr = data;

at the top and use `*ptr++` in the loop.

SenRamakri replied:
Good catch, fixed.

kjbracey commented on platform/mbed_error.c:
```diff
@@ -44,10 +47,46 @@ static core_util_atomic_flag error_in_progress = CORE_UTIL_ATOMIC_FLAG_INIT;
+//we dont have many uses cases to create a C wrapper for MbedCRC and the data
+//we calculate CRC on in this context is very less we will use a local 
+//implementation here.
+static unsigned int compute_crc32(void *data, int datalen)
```
data should be const

kjbracey reviewed (dismissed):
Approved, except for the CRC calculation error.

cmonr commented:
@SenRamakri The errors look legit. Unsure why they were only focused around the gnuarmeclipse exporter, and why the devices that failed seemed random.

One such error:
```
../mbed-os/platform/mbed_error.c: In function 'error':
Finished building: ../mbed-os/rtos/Thread.cpp
../mbed-os/platform/mbed_error.c:117:76: error: macro "ERROR_REPORT" passed 5 arguments, but takes just 4
         ERROR_REPORT(&last_error_ctx, "Fatal Run-time error", NULL, 0, true);
                                                                            ^
../mbed-os/platform/mbed_error.c:117:9: error: 'ERROR_REPORT' undeclared (first use in this function)
         ERROR_REPORT(&last_error_ctx, "Fatal Run-time error", NULL, 0, true);
         ^~~~~~~~~~~~
../mbed-os/platform/mbed_error.c:117:9: note: each undeclared identifier is reported only once for each function it appears in
../mbed-os/platform/mbed_error.c: In function 'mbed_error':
../mbed-os/platform/mbed_error.c:268:77: error: macro "ERROR_REPORT" passed 5 arguments, but takes just 4
         ERROR_REPORT(&last_error_ctx, error_msg, filename, line_number, true);
                                                                             ^
../mbed-os/platform/mbed_error.c:268:9: error: 'ERROR_REPORT' undeclared (first use in this function)
         ERROR_REPORT(&last_error_ctx, error_msg, filename, line_number, true);
         ^~~~~~~~~~~~
../mbed-os/platform/mbed_error.c:295:12: warning: function declared 'noreturn' has a 'return' statement
     return MBED_ERROR_FAILED_OPERATION;
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~
At top level:
../mbed-os/platform/mbed_error.c:424:20: warning: 'name_or_unnamed' defined but not used [-Wunused-function]
 static const char *name_or_unnamed(const char *name)
                    ^~~~~~~~~~~~~~~
../mbed-os/platform/mbed_error.c:69:21: warning: 'compute_crc32' defined but not used [-Wunused-function]
 static unsigned int compute_crc32(void *data, int datalen)
                     ^~~~~~~~~~~~~
make: *** [mbed-os/platform/mbed_error.o] Error 1
make: *** Waiting for unfinished jobs....
mbed-os/platform/subdir.mk:107: recipe for target 'mbed-os/platform/mbed_error.o' failed
```

SenRamakri committed:
Fix crc calculation error, code-style issues and other fixes

cmonr commented:
@SenRamakri Fyi, this still needs astyle work before it can be restarted.

SenRamakri commented:
@cmonr - It appears the astyle doesn't report all the code style issues in one shot, so I have to keep pushing new changes until everything is resolved, its a bit annoying.

cmonr commented:
@SenRamakri Interesting. Good to know.

@0xc0170 FYI about the current astyle test.

0xc0170 commented:
Please create an issue , we will look at it.

Have you tried to run the command locally ? Does it produce different result?

SenRamakri committed:
Removing support for printing error-report after reboot and other minor fixes

SenRamakri commented:
@kjbracey-arm - Please re-review. I have fixed the crc calculation error and have added other fixes.

One thing I have specifically in this commit is removing the optional-configuration to print error-report to the terminal on reboot(note that we still print the message that the system rebooted due to error, its only the printing of  error-report that is removed, everything else is as before). This is in response to some errors I'm seeing with this PR in CI. On debugging those I found that using mbed_error_printf() early during boot conflicts with current implementation of mbed_error_printf for IAR compiler as it ends up initializing c-lib mutexes but we do have a crit-sect in mbed_error_printf which causes the error. Although we can fix this, it takes changing mbed_error_printf() and or other functions which I think is not necessary to support this optional feature. If the user is connected to terminal the report will be printed to terminal before reboot anyway. So I have removed it for now. 
The test has also been updated to take care of some corner cases.  The design-document has also been updated to reflect this config change.

SenRamakri commented:
@0xc0170 - I have been running astyle locally. Im also seeing that when I run astyle locally it complains about max-indentation option. I have to remove that from .astylerc to run astyle locally. Is this because of version difference? What version I should be running locally?

0xc0170 commented:
> @0xc0170 - I have been running astyle locally. Im also seeing that when I run astyle locally it complains about max-indentation option. I have to remove that from .astylerc to run astyle locally. Is this because of version difference? What version I should be running locally?

AStyle v3.1. I've never seen that error (I assume you run older version as I previously tested 2.x versions and some did not provide options we have - they either changed or were added later).

Thanks for fixing this PR.

bulislaw commented on cmsis/TARGET_CORTEX_M/TOOLCHAIN_ARM/except.S:
```diff
@@ -80,7 +80,8 @@ Fault_Handler   PROC
 Fault_Handler_Continue
                 MOV      R12,R3
-                LDR      R1,=mbed_fault_context     
+                LDR      R3,=mbed_fault_context      
```
Ah i thought we agreed not to modify the cmsis files in mbed :( but to engage with the team to do it  properly. It's ok for now as the code is already modified but i'll rise an defect against that.

SenRamakri replied:
@bulislaw - Thanks for review and yes, writing a defect to investigate would be great. One thing to notice is that mbed_fault_handler.c/h is one of Mbed files and not from Cmsis, may be that helps. Originally I was planning to place these files under platform, but there has been concern around placing Cortex-M specific files under platform directory(under TARGET_CORTEX/TARGET_CORTEX_M), so moved them to Cmsis.

bulislaw commented:
I don't see the size change being reported, where can I find it? @ARMmbed/mbed-os-maintainers

OPpuolitaival commented:
@bulislaw size change is reported by part of ci-morph-test which is not executed yet

cmonr commented:
@bulislaw The size change report as it exists now is run after the Test CI job completes.
@kegilbert @studavekar can give more information.

kegilbert commented:
@bulislaw In addition to the morph test comment, the benchmark tests are not being run as of now due to the office migration (new office network is not fully setup). Everything else should be running as expected. Will have the benchmark tests running ASAP.

0xc0170 commented:
While we are waiting for finalizing the review (the benchmark would be nice to have here!), we run the build/exporters stage

/morph build

cmonr commented:
CI started.

0xc0170 commented:
From the latest job, these are the numbers:

```
INFO 11/22/2018 03:46:47 AM 
|-------------|-------|------|-------|------|-------|---------------|----------------|-------|-------|
| APPLICATION |  TEXT | DATA |  BSS  | HEAP | STACK | RESERVED_HEAP | RESERVED_STACK |  ROM  |  RAM  |
|-------------|-------|------|-------|------|-------|---------------|----------------|-------|-------|
|   ethernet  |  7.61 | 0.0  |  -0.1 | 0.0  |  0.0  |      0.04     |      0.0       |  7.4  | -0.09 |
|  filesystem | 10.52 | 0.0  | -0.77 | 0.0  |  0.0  |      0.03     |      0.0       | 10.12 | -0.14 |
|     wifi    |  9.73 | 0.0  | -0.53 | 0.0  |  0.0  |      0.03     |      0.0       |  9.38 | -0.33 |
|     rtos    | 17.34 | 0.0  | -0.49 | 0.0  |  0.0  |      0.03     |      0.0       | 16.31 |  -0.2 |
|   baseline  | 19.16 | 0.0  |  -0.8 | 0.0  |  0.0  |      0.03     |      0.0       | 17.91 | -0.52 |
|-------------|-------|------|-------|------|-------|---------------|----------------|-------|-------|
```

0xc0170 commented:
@bulislaw Ready for final approval?

bulislaw commented:
Are the numbers above bytes?

0xc0170 commented:
@OPpuolitaival  percentage ?

The important bit I missed is the status reports the numbers :

`jenkins-ci/dynamic-memory-usage — Success, RTOS ROM(+8106 bytes) RAM(-56 bytes)`

bulislaw commented:
well if it's 17% then we can't enable it by default...

SenRamakri commented:
@0xc0170 @bulislaw - I'm little bit confused about those numbers. When I did local builds for GCC-RELEASE, I get totally different results. +8K bytes shown here is inconceivable. was this ever part of roll-up?

0xc0170 commented:
@ARMmbed/mbed-os-test Can you answer above question ^^

@SenRamakri Please review the logs (it tests dynamic usage - running some apps).

From the docs: - 

> **jenkins-ci/dynamic-memory-usage** - Report dynamic memory use compared to the master branch.

@ARMmbed/mbed-os-test Please update the docs to describe the numbers more in detail

bulislaw reviewed (commented):
Code looks good, but I'd like us to get to the bottom of the memory reporting before approving.

0xc0170 commented:
I marked this as "ready for merge" but not merging yet. @OPpuolitaival Can you provide some numbers to backup earlier reports?

We experienced issues with dynamic reports, and it's currently disabled in the CI pipeline. Therefore I question this numbers as well. @bulislaw Please advise

bulislaw reviewed (approved):
Approved as the number provided by Senthil are sensible
